tests/listeners/test_listeners.py

```diff
@@ -0,0 +1,140 @@
+TASK_ID = "test_listener_task"
+EXECUTION_DATE = timezone.utcnow()
+
+
+hookimpl = pluggy.HookimplMarker("airflow")
+
+
+class CollectingListener(Listener):
+    def __init__(self):
+        self.states = []
+
+
+class PartialListener(CollectingListener):
+    @hookimpl
+    def on_task_instance_running(self, previous_state, task_instance, session):
+        self.states.append(State.RUNNING)
+
+
+class FullListener(PartialListener):
+    @hookimpl
+    def on_task_instance_success(self, previous_state, task_instance, session):
+        self.states.append(State.SUCCESS)
+
+    @hookimpl
+    def on_task_instance_failed(self, previous_state, task_instance, session):
+        self.states.append(State.FAILED)
+
+
+class ThrowingListener(Listener):
+    @hookimpl
+    def on_task_instance_running(self, previous_state, task_instance, session):
+        raise RuntimeError()
+
+
+@pytest.fixture(scope="module", autouse=True)
+def register_events():
+    register_task_instance_state_events()
+
+
+@pytest.fixture(autouse=True)
+def clean_listener_manager():
+    lm = get_listener_manager()
+    lm.clear()
+    yield
+    lm = get_listener_manager()
+    lm.clear()
+
+
+@provide_session
+def test_listener_gets_calls(create_task_instance, session=None):
+    lm = get_listener_manager()
+    listener = FullListener()
+    lm.add_listener(listener)
+
+    ti = create_task_instance(session=session, state=State.QUEUED)
+    ti.run()
+
+    assert len(listener.states) == 2
+    assert listener.states == [State.RUNNING, State.SUCCESS]
+
+
+@provide_session
+def test_listener_gets_only_subscribed_calls(create_task_instance, session=None):
+    lm = get_listener_manager()
+    listener = PartialListener()
+    lm.add_listener(listener)
+
+    ti = create_task_instance(session=session, state=State.QUEUED)
+    ti.run()
+
+    assert len(listener.states) == 1
+    assert listener.states == [State.RUNNING]
+
+
+@provide_session
+def test_listener_throws_exceptions(create_task_instance, session=None):
+    lm = get_listener_manager()
+    listener = ThrowingListener()
+    lm.add_listener(listener)
+
+    with pytest.raises(RuntimeError):
+        ti = create_task_instance(session=session, state=State.QUEUED)
+        ti.run()
+
+
+def test_listener_needs_to_subclass_listener():
+    lm = get_listener_manager()
+
+    class Dummy:
+        @hookimpl
+        def on_task_instance_running(self, previous_state, task_instance, session):
+            pass
+
+    lm.add_listener(Dummy())
+    assert not lm.has_listeners()
+
+
+@provide_session
+def test_listener_captures_failed_taskinstances(create_task_instance_of_operator, session=None):
+    lm = get_listener_manager()
+    listener = FullListener()
+    lm.add_listener(listener)
+
+    with pytest.raises(AirflowException):
+        ti = create_task_instance_of_operator(
+            BashOperator, dag_id=DAG_ID, execution_date=EXECUTION_DATE, task_id=TASK_ID, bash_command="exit 1"
+        )
+        ti.run()
```

Comment:
Hmmm, if we can avoid calling `ti.run()` here in this test that would be good. (It does _a lot_ and is relatively heavy weight approach, so if we can do something lighter here that would be preferable)

Reply:
I can do `_run_raw_task`.